widget/task_preview.lua

```diff
@@ -119,81 +152,108 @@ local function draw_widget(
+    args.forced_height = args.forced_height or dpi(200)
+    args.input_passthrough = args.input_passthrough or false
+
+    args.margin = args.margin or beautiful.task_preview_widget_margin or dpi(0)
+    args.shape = args.shape or beautiful.task_preview_widget_shape or nil
+    args.bg = args.bg or beautiful.task_preview_widget_bg or "#000000"
+    args.border_width = args.border_width or beautiful.task_preview_widget_border_width or nil
+    args.border_color = args.border_color or beautiful.task_preview_widget_border_color or "#ffffff"
+    args.image_shape = args.image_shape or beautiful.task_preview_image_shape or nil
+
+    local ret = gobject{}
+    ret._private = {}
+
+    gtable.crush(ret, task_preview)
+    gtable.crush(ret, args)
+
+    ret._private.widget = awful.popup
+    {
+        type = ret.type,
         visible = false,
         ontop = true,
-        placement = placement_fn,
-        widget = wibox.container.background, -- A dummy widget to make awful.popup not scream
-        input_passthrough = true,
+        placement = ret.placement,
+        input_passthrough = ret.input_passthrough,
         bg = "#00000000",
-    })
+        widget = wibox.container.background, -- A dummy widget to make awful.popup not scream
+    }
 
-    tag.connect_signal("property::selected", function(t)
+    capi.tag.connect_signal("property::selected", function(t)
         -- Awesome switches up tags on startup really fast it seems, probably depends on what rules you have set
         -- which can cause the c.content to not show the correct image
-        gears.timer
-        {
+        gtimer {
             timeout = 0.1,
             call_now  = false,
             autostart = true,
             single_shot = true,
             callback = function()
                 if t.selected == true then
                     for _, c in ipairs(t:clients()) do
-                        c.prev_content = gears.surface.duplicate_surface(c.content)
+                        c.prev_content = gsurface.duplicate_surface(c.content)
                     end
                 end
             end
         }
     end)
 
-    awesome.connect_signal("bling::task_preview::visibility", function(s, v, c)
-        if v then
-            -- Update task preview contents
-            task_preview_box.widget = draw_widget(
-                c,
-                opts.structure,
-                screen_radius,
-                widget_bg,
-                widget_border_color,
-                widget_border_width,
-                margin,
-                widget_width,
-                widget_height
-            )
-        else
-            task_preview_box.widget = nil
-            collectgarbage("collect")
-        end
+    return ret
 
-        if not placement_fn then
-            task_preview_box.x = s.geometry.x + widget_x
-            task_preview_box.y = s.geometry.y + widget_y
-        end
+    -- awesome.connect_signal("bling::task_preview::visibility", function(s, v, c)
```

Comment:
Might want to keep this for a few months so we can notify the user that this signal is deprecated.